docs/cryptodoc/src/05_09_kyber.rst

```diff
@@ -141,161 +302,115 @@ Based on these functions the key generation process follows **Algorithms 4 and 7
 
    **Notes:**
 
-   - ``H(pk)`` is precomputed in ``Kyber_PublicKeyInternal`` and accessible via ``H_public_key_bits_raw()``.
-   - The input/output structure corresponds to Botan's ``KEM_Encryption`` interface.
+   - Steps 1-3 correspond to Algorithms 20 and 17 of [FIPS-203]_ and are performed in :srcref:`[src/lib/pubkey/kyber]/ml_kem/ml_kem_impl.cpp:25|ML_KEM_Encryptor::encapsulate`.
+   - Steps 4-14 correspond to Algorithms 14 of [FIPS-203]_ and are performed in :srcref:`[src/lib/pubkey/kyber]/kyber/kyber_common/kyber_keys.cpp:55|indcpa_encrypt`.
+   - The transposed matrix ``At`` is precomputed and stored in the public key object.
 
 
+.. _pubkey/kyber/decaps:
+
 Key Decapsulation
 -----------------
 
-.. _pubkey/kyber/cpapke_dec:
-
-Kyber.CPAPKE
-^^^^^^^^^^^^
+The algorithms for high-level ML-KEM decapsulation and internal decapsulation
+(Algorithms 21 and 18 of [FIPS-203]_) are implemented in
+:srcref:`[src/lib/pubkey/kyber]/ml_kem/ml_kem_impl.cpp:48|ML_KEM_Decryptor::decapsulate`.
+They uses the K-PKE encapsulation and decapsulation algorithms (Algorithm 14
+and 15 of [FIPS-203]_) implemented in
+:srcref:`[src/lib/pubkey/kyber]/kyber/kyber_common/kyber_keys.cpp:55|Kyber_PublicKeyInternal::indcpa_encrypt`
+and
+:srcref:`[src/lib/pubkey/kyber]/kyber/kyber_common/kyber_keys.cpp:84|Kyber_PrivateKeyInternal::indcpa_decrypt`.
+In combination, Botan does the following:
 
-IND-CPA decryption works as follows, realizing **Algorithm 6** of [Kyber-R3]_:
-
-.. |step_3_formular| replace:: :math:`\mathbf{\hat{s}}^T \circ \mathsf{NTT}(\mathbf{u})`
-.. |step_4_formular| replace:: :math:`\mathsf{NTT}^{-1}(\mathbf{\hat{s}}^T \circ \mathsf{NTT}(\mathbf{u}))`
-.. |step_5_formular| replace:: :math:`v - \mathsf{NTT}^{-1}(\mathbf{\hat{s}}^T \circ \mathsf{NTT}(\mathbf{u}))`
-.. admonition:: Kyber_KEM_Decryptor::indcpa_dec()
+.. admonition:: ML_KEM_Decryptor::decapsulate
 
    **Input:**
 
-   -  ``sk = (sk_s, z)``: secret key
-   -  ``c``: ciphertext bytes
+   -  ``c``: ciphertext
+   -  ``sk = (seed.d, seed.z)``: secret key
+   -  ``pk``: public key
 
    **Output:**
 
-   -  ``m``: message bytes (decapsulated key)
+   -  ``K_prime``: shared secret key
 
    **Steps:**
 
-   1. Create a ``Ciphertext`` object ``ct`` by decoding and decompressing the ciphertext bytes. (L. 1-2, Alg. 6 [Kyber-R3]_)
-   2. ``ct.b.ntt()``
-   3. ``mp = sk_s * ct.b``  (|step_3_formular| of L. 4, Alg. 6 [Kyber-R3]_)
-   4. ``mp.invntt()`` (|step_4_formular| of L. 4, Alg. 6 [Kyber-R3]_)
-   5. ``mp -= ct.v`` (|step_5_formular| of L. 4, Alg. 6 [Kyber-R3]_)
-   6. ``m = mp.to_message()`` (L. 4, Alg. 6 [Kyber-R3]_)
-
-   **Notes:**
-
-   - The coefficients of ``mp`` are additively inverse to the specification. For the subsequent compression, however, only the distances of the coefficients to zero are relevant, which are the same in both cases.
+   1. Recompute the secret key value ``s`` from ``seed.d``
+   2. K-PKE decrypt ``c`` to obtain message ``m``
```

Comment:
```suggestion
   1. Recompute the secret key value ``s`` from ``seed.d`` using ``expand_keypair``
   2. K-PKE decrypt ``c`` to obtain message ``m_prime``
```

I think it makes sense to use `m_prime` here, because it is essentially just a decryption "candidate" until the re-encryption was successful. Also FIPS 203 does it like that. Same would be true for u and v, but that would make it harder to read than necessary. Absolutely feel free to omit this suggestion, as it isn't really consistent either!

Reply:
This was a mistake of me. Even the code says m_prime. Good catch :)

Reply:
i